src/common/utility.cpp

```diff
@@ -563,6 +563,53 @@ QString Utility::currentCpuArch()
 }
 #endif
 
+void Utility::selfRestart()
+{
+    qCInfo(lcUtility) << "Restarting application with PID" << qApp->applicationPid();
+
+#ifdef Q_OS_LINUX
+    // on Linux, we should always prefer to use execv, which will guarantee an instant restart and won't cause any
+    // false-positive results with KDSingleApplication as the application is guaranteed to have terminated before the
+    // re-launch
+    QString pathToLaunch = QCoreApplication::applicationFilePath();
+
+    if (Utility::runningInAppImage()) {
+        pathToLaunch = Utility::appImageLocation();
+    }
+
+    // absolute paths make things a lot easier typically!
+    pathToLaunch = QFileInfo(pathToLaunch).absoluteFilePath();
+
+    // must copy all arguments to mutable strings
+    // we don't care about the little bit of memory that could possibly be leaked, hence just using strdup
+    std::vector<char *> args;
+
+    // not strictly needed, but good style
+    if (Utility::runningInAppImage()) {
+        args.push_back(strdup(pathToLaunch.toStdString().c_str()));
+    } else {
+        args.push_back(strdup(QCoreApplication::arguments().front().toStdString().c_str()));
+    }
+
+    for (const auto &arg : QCoreApplication::arguments().sliced(1)) {
+        args.push_back(strdup(arg.toStdString().c_str()));
+    }
+
+    // must be null-terminated, as we don't pass a length
+    args.push_back(nullptr);
+
+    execvp(pathToLaunch.toStdString().c_str(), args.data());
```

Comment:
Well that might work with a simple application, but with this code the old instance was not properly shut down?

Reply:
Of course it is. `exec*` terminates the old application properly, i.e., all of Qt's shutdown code is run.